File: js/samples/08.datasource.azureAISearch/src/AzureAISearchDataSource.ts
import { DataSource, Memory, OpenAIEmbeddings, RenderedPromptSection, Tokenizer } from '@microsoft/teams-ai';
import { AzureKeyCredential, SearchClient, GeographyPoint } from '@azure/search-documents';
import { TurnContext } from 'botbuilder';

/**
 * Defines the Restaurant Interface.
 */
export interface Restaurant {
    restaurantId?: string;
    restaurantName?: string | null;
    description?: string | null;
    descriptionVectorEn?: number[] | null;
    category?: string | null;
    tags?: string[] | null;
    rating?: number | null;
    location?: GeographyPoint | null;
    address?: {
        streetAddress?: string | null;
        city?: string | null;
        stateProvince?: string | null;
        postalCode?: string | null;
        country?: string | null;
    } | null;
}

/**
 * Options for creating a `AzureAISearchDataSource`.
 */
export interface AzureAISearchDataSourceOptions {
    /**
     * Name of the data source. This is the name that will be used to reference the data source in the prompt template.
     */
    name: string;

    /**
     * Name of the Azure AI Search index.
     */
    indexName: string;

    /**
     * Azure OpenAI API key. AzureO
     */
    azureOpenAIApiKey: string;

    /**
     * Azure OpenAI endpoint. This is used to generate embeddings for the user's input.
     */
    azureOpenAIEndpoint: string;

    /**
     * Azure OpenAI Embedding deployment. This is used to generate embeddings for the user's input.
     */
    azureOpenAIEmbeddingDeployment: string;

    /**
     * Azure AI Search API key.
     */
    azureAISearchApiKey: string;

    /**
     * Azure AI Search endpoint.
     */
    azureAISearchEndpoint: string;
}

/**
 * A data source that uses a Azure AI Search index to inject text snippets into a prompt.
 */
export class AzureAISearchDataSource implements DataSource {
    /**
     * Name of the data source and local index.
     */
    public readonly name: string;

    /**
     * Options for creating the data source.
     */
    private readonly options: AzureAISearchDataSourceOptions;

    /**
     * Azure AI Search client.
     */
    private readonly searchClient: SearchClient<Restaurant>;

    /**
     * Creates a new `AzureAISearchDataSource` instance.
     * @param {AzureAISearchDataSourceOptions} options Options for creating the data source.
     */
    public constructor(options: AzureAISearchDataSourceOptions) {
        this.name = options.name;
        this.options = options;
        this.searchClient = new SearchClient<Restaurant>(
            options.azureAISearchEndpoint,
            options.indexName,
            new AzureKeyCredential(options.azureAISearchApiKey),
            {}
        );
    }

    /**
     * Renders the data source as a string of text.
     * @remarks
     * The returned output should be a string of text that will be injected into the prompt at render time.
     * @param {TurnContext} context Turn context for the current turn of conversation with the user.
     * @param _context
     * @param {Memory} memory An interface for accessing state values.
     * @param {Tokenizer} tokenizer Tokenizer to use when rendering the data source.
     * @param {number} maxTokens Maximum number of tokens allowed to be rendered.
     * @returns {Promise<RenderedPromptSection<string>>} A promise that resolves to the rendered data source.
     */
    public async renderData(
        _context: TurnContext,
        memory: Memory,
        tokenizer: Tokenizer,
        maxTokens: number
    ): Promise<RenderedPromptSection<string>> {
        const query: string = memory.getValue('temp.input');

        // If the user input is empty, don't add any text to the prompt.
        if (!query) {
            return { output: '', length: 0, tooLong: false };
        }

        const selectedFields = [
            'restaurantId',
            'restaurantName',
            'description',
            'category',
            'tags',
            'rating',
            'location',
            'link',
            'address'
        ];

        //// TEXT SEARCH ////
        // Get the restaurants with the query has high lexical relevance to category, tags, and restaurant name.
        const searchResults = await this.searchClient.search(query, {
            searchFields: ['category', 'tags', 'restaurantName'],
            select: selectedFields as any
        });

        //// VECTOR SEARCH ////
        //// Get the restaurants with description that is the most similar to the user input.
        // const queryVector: number[] = await this.getEmbeddingVector(query);
        // const searchResults = await this.searchClient.search('*', {
        //     vectorSearchOptions: {
        //         queries: [
        //             {
        //                 kind: 'vector',
        //                 fields: ['descriptionVectorEn'],
        //                 kNearestNeighborsCount: 3,
        //                 // The query vector is the embedding of the user's input
        //                 vector: queryVector
        //             }
        //         ]
        //     },
        //     select: selectedFields as any
        // });

        //// HYBRID SEARCH ////
        // Search using both vector and text search
        // const queryVector: number[] = await this.getEmbeddingVector(query);
        // const searchResults = await this.searchClient.search(query, {
        //     searchFields: ['category', 'tags', 'restaurantName'],
        //     select: selectedFields as any,
        //     vectorSearchOptions: {
        //         queries: [
        //             {
        //                 kind: 'vector',
        //                 fields: ['descriptionVectorEn'],
        //                 kNearestNeighborsCount: 3,
        //                 // The query vector is the embedding of the user's input
        //                 vector: queryVector
        //             }
        //         ]
        //     },
        // });

        // Show example for how to make multiple search types

        if (!searchResults.results) {
            return { output: '', length: 0, tooLong: false };
        }

        // Concatenate the restaurant documents (i.e json object) string into a single document
        // until the maximum token limit is reached. This can be specified in the prompt template.
        let usedTokens = 0;
        let doc = '';
        for await (const result of searchResults.results) {
            const formattedResult = this.formatDocument(result.document);
            const tokens = tokenizer.encode(formattedResult).length;

            if (usedTokens + tokens > maxTokens) {
                break;
            }

            doc += formattedResult;
            usedTokens += tokens;
        }

        return { output: doc, length: usedTokens, tooLong: usedTokens > maxTokens };
    }

    /**
     * Formats the restaurant document as a json string .
     * @param {Restaurant} result The restaurant document to format.
     * @returns {string} The formatted restaurant document as a json string.
     */
    private formatDocument(result: Restaurant): string {
        return `<context>${JSON.stringify(result)}</context>`;
    }

    /**
     * Uses Azure OpenAI to generate embeddings for the user's input.
     * @param text The user's input.
     * @returns The embedding vector for the user's input.
     */
    private async getEmbeddingVector(text: string): Promise<number[]> {
        const embeddings = new OpenAIEmbeddings({
            azureApiKey: this.options.azureOpenAIApiKey,
            azureEndpoint: this.options.azureOpenAIEndpoint,
            azureDeployment: this.options.azureOpenAIEmbeddingDeployment
        });

        const result = await embeddings.createEmbeddings(this.options.azureOpenAIEmbeddingDeployment, text);
        if (result.status !== 'success' || !result.output) {
            throw new Error(`Failed to generate embeddings for description: ${text}`);
        }

        return result.output[0];
    }
}

File: dotnet/samples/08.datasource.azureaisearch/AzureAISearchBot/AzureAISearchDataSource.cs
using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;
using AzureAISearchIndexer;
using Microsoft.Bot.Builder;
using Microsoft.Teams.AI.AI.DataSources;
using Microsoft.Teams.AI.AI.Embeddings;
using Microsoft.Teams.AI.AI.Prompts.Sections;
using Microsoft.Teams.AI.AI.Tokenizers;
using Microsoft.Teams.AI.State;
using System.Text;

namespace AzureAISearchBot
{
    public class AzureAISearchDataSource : IDataSource
    {
        public string Name { get; }

        public readonly AzureAISearchDataSourceOptions Options;

        public readonly SearchClient SearchClient;

        public AzureAISearchDataSource(AzureAISearchDataSourceOptions options)
        {
            Options = options;
            Name = options.Name;

            AzureKeyCredential credential = new AzureKeyCredential(options.AzureAISearchApiKey);
            SearchClient = new SearchClient(options.AzureAISearchEndpoint, options.IndexName, credential);
        }

#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
        public async Task<RenderedPromptSection<string>> RenderDataAsync(ITurnContext context, IMemory memory, ITokenizer tokenizer, int maxTokens, CancellationToken cancellationToken = default)
#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
        {
            string query = (string)memory.GetValue("temp.input")!;

            if (string.IsNullOrEmpty(query))
            {
                return new RenderedPromptSection<string>("");
            }

            List<string> selectedFields = new() { "RestaurantId", "RestaurantName", "Description", "Category", "Tags", "Rating", "Address" };
            List<string> searchFields = new() { "Category", "Tags", "RestaurantName" };

            //// TEXT SEARCH ////
            //// Get the restaurants with the query has high lexical relevance to category, tags, and restaurant name.
            SearchOptions options = new();
            foreach (string field in searchFields)
            {
                options.SearchFields.Add(field);
            }

            foreach (string field in selectedFields)
            {
                options.Select.Add(field);
            }
            SearchResults<Restaurant> search = SearchClient.Search<Restaurant>(query, options);

            //// VECTOR SEARCH ////
            //// Get the restaurants with description that is the most similar to the user input.
            /*SearchOptions options = new();
            ReadOnlyMemory<float> vectorizedQuery = await this._GetEmbeddingVector(query);
            foreach (string field in selectedFields)
            {
                options.Select.Add(field);
            }
            options.VectorSearch = new()
            {
                Queries = { new VectorizedQuery(vectorizedQuery) { KNearestNeighborsCount = 3, Fields = { "DescriptionVectorEn" } } }
            };
            SearchResults<Restaurant> search = SearchClient.Search<Restaurant>(options);
            */

            //// HYBRID SEARCH ////
            //// Search using both vector and text search
            /*SearchOptions options = new();
            ReadOnlyMemory<float> vectorizedQuery = await this._GetEmbeddingVector(query);
            foreach (string field in searchFields)
            {
                options.SearchFields.Add(field);
            }

            foreach (string field in selectedFields)
            {
                options.Select.Add(field);
            }
            options.VectorSearch = new()
            {
                Queries = { new VectorizedQuery(vectorizedQuery) { KNearestNeighborsCount = 3, Fields = { "DescriptionVectorEn" } } }
            };
            SearchResults<Restaurant> search = SearchClient.Search<Restaurant>(query, options);
            */

            // Concatenate the restaurant documents (i.e json object) string into a single document
            // until the maximum token limit is reached. This can be specified in the prompt template.
            int usedTokens = tokenizer.Encode("Contexts: ").Count;
            StringBuilder doc = new StringBuilder("Contexts: ");
            Pageable<SearchResult<Restaurant>> results = search.GetResults();
            foreach (SearchResult<Restaurant> result in results)
            {
                string document = $"<context>{result.Document}</context>";
                int tokens = tokenizer.Encode(document).Count;

                if (usedTokens + tokens > maxTokens)
                {
                    break;
                }

                doc.Append(document);
                usedTokens += tokens;
            }

            return new RenderedPromptSection<string>(doc.ToString(), usedTokens, usedTokens > maxTokens);
        }

#pragma warning disable IDE0051 // Remove unused private members
        private async Task<ReadOnlyMemory<float>> _GetEmbeddingVector(string query)
#pragma warning restore IDE0051 // Remove unused private members
        {
            AzureOpenAIEmbeddingsOptions options = new(this.Options.AzureOpenAIApiKey, this.Options.AzureOpenAIEmbeddingDeployment, this.Options.AzureOpenAIEndpoint);
            OpenAIEmbeddings embeddings = new(options);
            EmbeddingsResponse response = await embeddings.CreateEmbeddingsAsync(new List<string> { query });

            return response.Output!.First();
        }
    }

    public class AzureAISearchDataSourceOptions
    {
        /// <summary>
        /// Name of the data source
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Name of the Azure AI Search index
        /// </summary>
        public string IndexName { get; set; }

        /// <summary>
        /// Azure OpenAI API key
        /// </summary>
        public string AzureOpenAIApiKey { get; set; }

        /// <summary>
        /// Azure OpenAI endpoint
        /// </summary>
        public string AzureOpenAIEndpoint { get; set; }

        /// <summary>
        /// Azure AI Search API key
        /// </summary>
        public string AzureAISearchApiKey { get; set; }

        /// <summary>
        /// Azure AI Search endpoint
        /// </summary>
        public Uri AzureAISearchEndpoint { get; set; }

        /// <summary>
        /// Azure OpenAI embeddings deployment name
        /// </summary>
        public string AzureOpenAIEmbeddingDeployment { get; set; }
    }
}
